### PR TITLE
[FEM] Add method computeVonMisesStress in TetrahedralCorotationalForceField and option to draw them

### PR DESCRIPTION
**Pull request by epernod**
Similar to TetrahedronFEMForceField, compute von mises per element and add option to draw it.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.h [hunk 1]
```diff
@@ -190,12 +192,22 @@ class TetrahedralCorotationalFEMForceField : public BaseLinearElasticityFEMForce
     Data<sofa::type::RGBAColor> d_drawColor3; ///<  draw color for faces 3
     Data<sofa::type::RGBAColor> d_drawColor4; ///<  draw color for faces 4
     Data<std::map < std::string, sofa::type::vector<double> > > _volumeGraph;
+    
+    Data<bool> d_computeVonMisesStress;
+    Data<type::vector<Real> > d_vonMisesPerElement; ///< von Mises Stress per element
+    Data<type::vector<Real> > d_vonMisesPerNode; ///< von Mises Stress per node
+    
+    sofa::helper::ColorMap* m_VonMisesColorMap;
```
Comment (fredroy) [inline Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.h]:
```suggestion
    Data<type::vector<Real> > d_vonMisesPerNode; ///< von Mises Stress per node
    
    sofa::helper::ColorMap* m_vonMisesColorMap;
```
I guess this is a copy-paste from the other file 🫢but it is nice to have consistency

File: Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl [hunk 2]
```diff
@@ -874,7 +894,19 @@ void TetrahedralCorotationalFEMForceField<DataTypes>::initLarge(int i, Index&a,
...
+    type::Mat<4, 4, Real> matVert;
+    const core::topology::BaseMeshTopology::Tetrahedron tetra = { a, b, c, d };
+    for (Index k = 0; k < 4; k++) {
+        Index ix = tetra[k];
+        matVert[k][0] = 1.0;
+        for (Index l = 1; l < 4; l++)
+            matVert[k][l] = X0[ix][l - 1];
+    }
+
+    const bool canInvert = type::invertMatrix(tetrahedronInf[i].elemShapeFun, matVert);
```
Comment (fredroy) [inline Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl]:
```suggestion
    [[maybe_unused]] const bool canInvert = type::invertMatrix(tetrahedronInf[i].elemShapeFun, matVert);
```

File: Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl [hunk 3]
```diff
@@ -1197,12 +1229,119 @@ void TetrahedralCorotationalFEMForceField<DataTypes>::computeBBox(const core::Ex
...
+        type::Vec<6, Real> vStrain;
+        type::Mat<3, 3, Real> gradU;
+        Transformation R_0_2;
+        Displacement D;
+
+        const auto& tetra = tetras[i];
+        const TetrahedronInformation& tetraInfo = tetrahedronInf[i];
+
+        computeRotationLarge(R_0_2, X, tetra[0], tetra[1], tetra[2]);
+        //tetraInfo.rotation.transpose(R_0_2);
```
Comment (fredroy) [inline Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl]:
```suggestion
```

File: Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl [hunk 4]
```diff
@@ -1211,46 +1350,84 @@ void TetrahedralCorotationalFEMForceField<DataTypes>::draw(const core::visual::V
...
+                pts[nd] = x[nd];
+                nodeColors[nd] = evalColor(vMN[nd]);
+            }
+            vparams->drawTool()->drawPoints(pts, 10, nodeColors);
+        }
     }
+    
+    if (d_drawing.getValue())
+    {
+        std::vector< type::Vec3 > points[4];
```
Comment (fredroy) [inline Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl:1388]: You could reserve the memory beforehand as the size seems computable

**Review by fredroy — CHANGES_REQUESTED** [review]

**Comment by damienmarchal** [comment]
To me adding more and more hard coded drawing function by copy-pasting other's drawing function in every components is not desirable. 

I see two better solutions: 
1) the drawing code & related datafield should at least be moved in a separated c++ object to factor rendering & UX. 
2)  the drawing code & related datafields should be in a separated sofa component so they can be compose in the scene.

**Review by bakpaul — APPROVED** [review]

File: Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl (hunk 4 above)
Comment (epernod) [reply Sofa/Component/SolidMechanics/FEM/Elastic/src/sofa/component/solidmechanics/fem/elastic/TetrahedralCorotationalFEMForceField.inl:1388]: I did it but in a new PR I will change that drawing to use directly the drawScaledTetrahedra from drawTools

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Review by fredroy — APPROVED** [review]

**Pull request merged**